Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -740,7 +740,7 @@ public final class PackageBuilder {
 
         // Compute the path to public headers directory.
         let publicHeaderComponent = manifestTarget.publicHeadersPath ?? ClangTarget.defaultPublicHeadersComponent
-        let publicHeadersPath = potentialModule.path.appending(try RelativePath(validating: publicHeaderComponent))
+        let publicHeadersPath = try AbsolutePath(RelativePath(validating: publicHeaderComponent), relativeTo: potentialModule.path)
```

Comment:
This worries me slightly - the `RelativePath(validating:)` can still truncate the path if it references unanchored ancestors.

Reply:
Is the worry that this would be a regression or that it would be an incomplete fix to the problem?  Re: Ted's question in https://github.com/apple/swift-package-manager/pull/4279

Reply:
I see, so you're referring to this thing:  https://github.com/apple/swift-tools-support-core/pull/305#pullrequestreview-922466642

With the removal of AbsolutePath API, what is then the right way nowadays to safely append a path represented as a relative string (possibly having ".." references) relative to an absolute path?

Reply:
There is `public init(_ absPath: AbsolutePath, _ relStr: String)` but the implementation of that just turns around and converts the string to a `RelativePath`, which is the one that causes concern.

Reply:
This (having a known absolute path and wanting to add a relative path to it) is a fairly common use case that it seems to me we should end up with a straightforward way to handle in a way that does the right normalization.

Reply:
personally I feel the deprecation of `AbsolutePath::appending(RelativePath)` is an API design regression - its a nice way to express the concatenation of path compared to the initializer based approach. @compnerd is there a way to un-deprecate that method and make it do the right thing for Windows instead? it will make this PR and the underlying TSC one unnecessary

Reply:
I'm starting to wonder the same thing.  I was convinced by the original reasoning that the normalization needs to be done on the whole string and not just each piece before they're concatenated together, and that part I agree with, but as Tom said it's become hard to do things that feel like they should be simple.

Also, I think there was a comment that `RelativePath` could end up truncating leading `..` (did I understand that comment correct — that doesn't seem right to me, as there needs to be a way to represent a relative path that goes upward in the path).  Could we defer the processing of that until it gets appended to an absolute path?

Reply:
BTW, I find the behavior of `appending` w.r.t. to `..` a bit surprising in the first place, e.g. if you do `AbsolutePath.root.appending(components: ["a", "..", "b"])`, you end up with `/b`. Maybe this kind of normalization should generally be an optional secondary step?

Reply:
@compnerd wdyt?

Reply:
Sorry about the delay, I didn't get a notification for some reason.

I think that Anders' and I see things somewhat similarly.  The original problem stands and will continue to cause trouble.  Delaying the path normalization would resolve the issue because it dealing with the underlying issue: we need to perform the normalization post concatenation.  I think that the current behaviour of the normalization can be surprising.  I think that restoring the API and delaying the normalization (perhaps as an explicit step) would be my preference - but I cannot definitely say if that is the interface we want to expose to users.  If everyone agrees that is the right API design choice, that would allay my issues with the approach.

Reply:
>  I think that restoring the API and delaying the normalization (perhaps as an explicit step) would be my preference - but I cannot definitely say if that is the interface we want to expose to users. If everyone agrees that is the right API design choice, that would allay my issues with the approach.

this sounds good to me I think. what are the user facing implications of delaying the normalization?

Reply:
Just to make sure we're talking about the same thing, would the normalization be delayed:
- until the relative path is applied to an absolute path?
- until the absolute path is "used" in some way (e.g. in a FileSystem operation or until a pathString is asked for)?
 
If we delay until the path is used, then code that compares paths as equal today could start failing.  But if we just delay normalization of relative paths until they are concatenated to an absolute path, then that would make sense to me.  Relative paths are never really "used" on their own in our model, and so it makes sense to not try to normalize them.

> BTW, I find the behavior of appending w.r.t. to .. a bit surprising in the first place, e.g. if you do AbsolutePath.root.appending(components: ["a", "..", "b"]), you end up with /b. Maybe this kind of normalization should generally be an optional secondary step?

I think the behaviour makes sense to me here but the word "appending" matches poorly with what happens.  It's more natural for this normalization to happen when we think of this as "making an absolute path by basing a relative path in some other absolute path".

Reply:
@compnerd what is the next step here? we would like to address the warnings this deprecation created ASAP

Reply:
@compnerd could you please help restore (un-deprecate) `AbsolutePath::appending(RelativePath)` and update it to do the right thing? this way this PR (and its TSC counterpart) is not needed as the warning will go away in a cleaner way